packages/actor-abstract-path/lib/ActorAbstractPath.ts

```diff
@@ -107,4 +108,43 @@ export abstract class ActorAbstractPath extends ActorQueryOperationTypedMediated
       }
     });
   }
+
+  public async ALPTwoVariables(subjectString: string, objectString: string, subjectVal: Term,
+    x: Term, predicate: Algebra.PropertyPathSymbol, context: ActionContext,
+    termHashes: {[id: string]: Term}, it: BufferedIterator<Bindings>, counter: any): Promise<void> {
+    const termString = termToString(subjectVal) + termToString(x);
+    if (termHashes[termString]) {
+      return;
+    }
+
+    (<any> it)._push(Bindings({ [subjectString]: subjectVal, [objectString]: x }));
+    termHashes[termString] = x;
+    counter.count++;
+
+    const thisVariable = this.generateVariable();
+    const vString = termToString(thisVariable);
+    const path = ActorAbstractPath.FACTORY.createPath(x, predicate, thisVariable);
+    const results = ActorQueryOperation.getSafeBindings(
+      await this.mediatorQueryOperation.mediate({ operation: path, context }),
+    );
+    results.bindingsStream.on('data', async bindings => {
+      const result = bindings.get(vString);
+      await this.ALPTwoVariables(
+        subjectString,
+        objectString,
+        subjectVal,
```

Comment:
No, it's still subjectVal. Let's say the function started with as subjectVal :a. The query is ?s :p+ ?o, and you found after one step for example :b, you return a binding with ?s :a and ?o :b. Then you start looking one step further (starting with result (as 4th parameter)). If you get then :c, you want to return ?s :a and ?o :c, so we still need the :a, which is the subjectVal we got in the first place.

Reply:
Simpler said: we want a reference to the ?s this all started with, so this does not need to shift along the way

Reply:
Ah yes, I misread and thought that one was also used for creating the followup path. But is the termString generated correct then? If your dataset is `:a :p :b. :b :p :c.` and you start with the `:a` your termHashes are going to be `:a:b` and `:a:c`. If the call to this function is then done with `:b` it is still going to run since it doesn't see the termhash `:b:c`, right?

Reply:
Note that I also realize the binding with subjectVal is necessary so I'm not immediately sure what the solution would be if I'm right.

Reply:
It is gonna run, that's true. But indeed for another subject value.

Reply:
So at this point, does this function actually work differently from the original ALP function besides adding the bindings of the subject inside of the functions instead of outside of the function as was done previously?

Reply:
Could this be a solution:
The termstring in ALPTwoVariables would be the object. So in your case first the b. We set it the termHashes on []. And then we call normal ALP from b and we listen to dataevents and then add it to that []. When the second b comes in ALPTwoVariables, we check if termstring exists and if so push our new subject with all the objects in the list.

Reply:
There is only one problem with this I think. If the end event has not yet occured on the ALP run on b and a second b comes in as object in ALPTwoVariables, the list may not be complete...

Reply:
> So at this point, does this function actually work differently from the original ALP function besides adding the bindings of the subject inside of the functions instead of outside of the function as was done previously?

I does. When :a :b en :z :b are both solutions, this is now possible because of the changed termhashes, the :z :b would not be a solution before, because the :b would be in the termhashes and it wouldn't be emitted for the z. But this can be fixed by not sharing the termshashes anymore I see now.

Reply:
Yes that's what I meant, this solution is the same as the previous solution (where the termHashes weren't shared). But I do think there is potential in this solution somewhere to change that.

Reply:
Not again btw that I could totally be wrong and there actually isn't a feasible way to do this more efficient.